src/relay.cpp

```diff
@@ -0,0 +1,196 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <rclcpp/generic_publisher.hpp>
+#include <rclcpp/generic_subscription.hpp>
+
+#include <memory>
+#include <string>
+
+using namespace std::chrono_literals;
+
+const auto LOOP_PERIOD = 100ms;
+
+namespace topic_tools
+{
+
+  class RelayNode : public rclcpp::Node
+  {
+  public:
+    RelayNode(
+      const std::string node_name, 
+      const rclcpp::NodeOptions &options);
+      
+    RelayNode(
+      const std::string node_name, 
+      const std::string input_topic, 
+      const std::string output_topic);
+
+  private:
+    void incoming_message_callback(std::shared_ptr<rclcpp::SerializedMessage> msg);
+    void timer_callback();
+    void subscribe();
+    void unsubscribe();
+    void initialize();
+    std::optional<std::string> try_find_topic_type();
+
+    rclcpp::GenericSubscription::SharedPtr sub_;
+    rclcpp::GenericPublisher::SharedPtr pub_;
+    rclcpp::TimerBase::SharedPtr timer_;
+    std::string input_topic;
+    std::string output_topic;
+    bool lazy;
+    std::optional<std::string> topic_type;
+  };
+
+  RelayNode::RelayNode(
+    const std::string node_name, const rclcpp::NodeOptions &options = rclcpp::NodeOptions()) 
+    : rclcpp::Node(node_name, options)
+  {
+    input_topic = declare_parameter<std::string>("input_topic");
+    output_topic = declare_parameter<std::string>("output_topic", input_topic + "_relay");
+    initialize();
+  }
+
+  RelayNode::RelayNode(
+      std::string node_name,
+      std::string input_topic,
+      std::string output_topic) : rclcpp::Node(node_name),
+                                  input_topic(input_topic),
+                                  output_topic(output_topic)
+  {
+    initialize();
+  }
+
+  void RelayNode::initialize()
+  {
+    lazy = declare_parameter<bool>("lazy", false);
+
+    auto ros_clock = rclcpp::Clock::make_shared();
+    timer_ = rclcpp::create_timer(this, ros_clock, LOOP_PERIOD, [=]()
+                                  { timer_callback(); });
+
+    // so that we execute the loop the first time instantly
+    timer_callback();
+  }
+
+  void RelayNode::timer_callback()
+  {
+    if (!topic_type)
+    {
+      if (topic_type = try_find_topic_type())
+      {
+        // we just found the topic type, so create the publisher and the subscriber
+        pub_ = this->create_generic_publisher(output_topic, *topic_type, rclcpp::QoS(1));
```

Comment:
Using the default QoS isn't going to cut it here - this is the one major difference for ROS 2 that we need to be careful of. We want to mimick the incoming data source as accurately as possible, for reasons of both compatibility and behavior.

In the best case scenario, there is only one source (or multiple sources that all have effectively the same QoS profile). I'd say that we can do pretty good most of the time by choosing the QoS of the first publisher we find.

To put all that mumbo-jumbo into some code, if we find a data source out there, we can:

```
auto publishers_info = this->get_publishers_info_by_topic(topic_name);
std::optional<rclcpp::QoS> qos_profile;
if (!publishers_info.empty()) {
  qos_profile = publishers_info[0].qos_profile();
} 
// Now you can use that qos_profile for the relay-publisher, to mimick at least the first one we found perfectly
```

That makes it difficult for us to create this publisher "ahead of time" to allow for laziness - we have no idea what potential destinations will actually be compatible with the QoS of the source that we end up relaying.

Instead of relying on `get_subscrition_count` to check if there are destinations out there looking for our relay topic, instead we should be able to use `this->get_subscriptions_info_by_topic` to count whether any subscribers exist, and only create the relay-subcriber if both a source exists and a destination exists.

Hopefully that makes some sense, I can help clarify more if needed.